cmp/options.go

```diff
@@ -396,68 +396,68 @@ func (visibleStructs) filter(_ *state, _ reflect.Type, _, _ reflect.Value) appli
 	panic("not implemented")
 }
 
-// reportFlags is a bit-set representing how a comparison was determined.
-type reportFlags uint
+// ReportFlags is a bit-set representing how a comparison was determined.
+type ReportFlags uint
 
 const (
-	_ reportFlags = (1 << iota) / 2
-
-	// reportEqual reports whether the node is equal.
-	// This may be ORed with reportByMethod or reportByFunc.
-	reportEqual
-	// reportUnequal reports whether the node is not equal.
-	// This may be ORed with reportByMethod or reportByFunc.
-	reportUnequal
-	// reportIgnored reports whether the node was ignored.
-	reportIgnored
-
-	// reportByMethod reports whether equality was determined by calling the
-	// Equal method. This may be ORed with reportEqual or reportUnequal.
-	reportByMethod
-	// reportByFunc reports whether equality was determined by calling a custom
-	// Comparer function. This may be ORed with reportEqual or reportUnequal.
-	reportByFunc
+	_ ReportFlags = (1 << iota) / 2
+
+	// ReportEqual reports whether the node is equal.
+	// This may be ORed with ReportByMethod or ReportByFunc.
+	ReportEqual
```

Comment:
Can this ever not be ORed with anything? If so, it'll be pretty easy to misuse.

Instead of exposing the implementation detail of `ReportFlags` being a bitset in the API, how about giving it `Equal`, `Unequal`, `Ignored`, `ByMethod`, and `ByFunc` methods instead?

Reply:
The downside to individual methods is that we cant add new ones to the reporter interface. For example, when we add cycle-detection support (#85), we would probably want to add `ReportByCycle` to indicate that the comparison was determined because of a detected cycle.

What are your thoughts on an opaque bit-set type?
```go
type ReportFlags struct {
    _     [0]func() // Make ReportFlags incomparable
    flags uint
}

func (f ReportFlags) Has(f2 ReportFlags) bool {
    return f&f2 != 0
}

var (
    ReportEqual    = ReportFlags{flags: 1}
    ReportUnequal  = ReportFlags{flags: 2}
    ReportIgnored  = ReportFlags{flags: 4}
    ReportByMethod = ReportFlags{flags: 8}
    ReportByFunc   = ReportFlags{flags: 16}
)
```

This way, it is impossible to accidentally do `==` on the flags. The only operation that you can do is call the `Has` method.

Reply:
The downside is that `ReportEqual` is not a variable instead of a constant. In theory someone could over-write it, but thats true for any sentinel variable like `io.EOF`.

Reply:
I'm suggesting methods on the `ReportFlags` type (and perhaps rename it to `Report` in case something non-flag is added to it in the future):

```go
func (Reporter) Report(r Report) {
  if !r.Equal() {
    // ...
  }
}
```

Having both `Equal` and `Unequal` methods looks a bit strange, so perhaps either define `Equal` as returning true for ignored values or make Equal/Unequal/Ignored an enum:

```go
if r.Type() == cmp.ReportEqual {
}
```

Reply:
I like the suggestion. Take a look.

Reply:
I like it.